### ISSUE TITLE
Sourcemap support?

### ISSUE THREAD

**Issue by rexxars** [opbeat/opbeat-node#74]
I have not looked into how the error reporting actually functions under the hood, so I'm not sure this is even possible, but would it be possible to add sourcemap support?

Basically, if I use Babel to compile `src/server.js` to `lib/server.js` with sourcemaps enabled, is there any way it could report the errors with the correct location?

**Comment by watson** [comment]
@rexxars it's on our roadmap and we have a feature coming up - hopefully next week - which might make it irrelevant - at least for coffee script, but maybe also for Babel. Do you know if Babel overwrites the normal error stack traces? If that's the case we might be able to use the overwritten stack trace as a source instead of a sourcemap.

**Comment by rexxars** [comment]
Sweet. I'm not sure if it overwrites the stack trace, sorry.

**Comment by hanjukim** [comment]
I'm running server with `babel-register` and when Error was caught stack traces shows correct line, so I guess babel does overwrites the Error stack traces.

**Comment by rexxars** [comment]
I'm compiling with [babel-cli](https://babeljs.io/docs/usage/cli/#compile-with-source-maps), in this case sourcemaps doesn't seem to have any effect for opbeat. 

It might be that a library such as [source-map-support](https://github.com/evanw/node-source-map-support) would fix this, but I'm wondering if perhaps it's better for `opbeat-node` to try to figure out sourcemaps using the [source-map](https://github.com/mozilla/source-map) module instead?

**Comment by watson** [comment]
@rexxars no I think you're right. It is something we'd like to support natively. Would it be possible for you to use the `retainLines` option of `babel-reguster` as @hanjukim mentioned for now to work around it?

```js
const register = require('babel-register')
register({ retainLines: true })
```

**Comment by rexxars** [comment]
No big deal right now, we can usually figure it out for ourselves.
Don't think we want to run `babel-register` in production.

**Comment by hfjallemark** [comment]
@rexxars any update on the feature that was released that would make this irrelevant? Would be great with native source map support!

**Comment by rexxars** [comment]
That's a question for @watson :-)

**Comment by hfjallemark** [comment]
My fault @rexxars, tagged the wrong guy! @watson, any update?

**Comment by watson** [comment]
@hfjallemark Unfortunately no. At the moment I don't have a good solution for Babel besides the workaround I mentioned above :(

**Comment by rexxars** [comment]
@watson Would opbeat accept a pull request that added sourcemap support using the [source-map](https://github.com/mozilla/source-map) module?

**Comment by watson** [comment]
@rexxars We'd love the help, but I'm not really sure how you'd go about it? I guess you want to convert the stack trace prior to sending it to the Opbeat intake API, or? Could you try and outline what approach you'd use first?

**Comment by rexxars** [comment]
Have not had a deep dive into the internals of this, so I might be underestimating it.

I was thinking of basically doing the same as the [source-map-support](https://github.com/evanw/node-source-map-support) module, only instead of registering handlers to read and parse the sourcemaps on uncaught exception events, I could hook it up to the existing opbeat agent where it handles errors.

Not sure if I'll ever get the time to do it, but sounds like a fun challenge.

**Comment by rexxars** [comment]
I did some experimenting in this field - slightly harder than I expected it to be due to abstractions, which makes it harder to find the proper API and ways to hook in. `opbeat-node` uses stackman, which I [sent a PR to](https://github.com/watson/stackman/pull/4) and never heard back, so it's hard for me to get any further on this without some feedback.

**Comment by watson** [comment]
@rexxars thanks, I'll take a look at the stackman PR

**Comment by watson** [comment]
The source map support will be included in the next minor version of the Opbeat agent. It's scheduled for release on Monday.

**Comment by watson** [comment]
We just released version 4.13.0 of the agent that includes the source map support. Read all about how to leverage it here: https://opbeat.com/blog/posts/announcing-source-map-support-in-nodejs/

**Comment by rexxars** [comment]
:tada: :+1:
